staging/src/k8s.io/kms/internal/plugins/_mock/pkcs11/pkcs11.go

```diff
@@ -0,0 +1,123 @@
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package pkcs11
+
+import (
+	"context"
+	"crypto/rand"
+	"fmt"
+
+	crypot11 "github.com/ThalesIgnite/crypto11"
+
+	"k8s.io/kms/pkg/service"
+	"k8s.io/kms/pkg/value"
+)
+
+const (
+	mockAnnotationKey = "version.encryption.remote.io"
+)
+
+var _ service.Service = &pkcs11RemoteService{}
+
+type pkcs11RemoteService struct {
+	key     *crypot11.SecretKey
+	keyID   string
+	dataCtx value.DefaultContext
+}
+
+// NewPKCS11RemoteService creates a new PKCS11 remote service with SoftHSMv2 configuration file and keyID
+func NewPKCS11RemoteService(configFilePath, keyID string) (service.Service, error) {
+	ctx, err := crypot11.ConfigureFromFile(configFilePath)
+	if err != nil {
+		return nil, err
+	}
+
+	if len(keyID) == 0 {
+		return nil, fmt.Errorf("invalid keyID")
+	}
+
+	remoteService := &pkcs11RemoteService{
+		keyID: keyID,
+	}
+
+	if remoteService.key, err = ctx.FindKey(nil, []byte(keyID)); err != nil {
+		return nil, err
+	}
+	if remoteService.key == nil {
+		return nil, fmt.Errorf("key not found")
+	}
+
+	return remoteService, nil
+}
+
+func (s *pkcs11RemoteService) Encrypt(ctx context.Context, uid string, plaintext []byte) (*service.EncryptResponse, error) {
+	aead, err := s.key.NewGCM()
+	if err != nil {
+		return nil, err
+	}
+	nonceSize := aead.NonceSize()
+	result := make([]byte, nonceSize+aead.Overhead()+len(plaintext))
+	n, err := rand.Read(result[:nonceSize])
+	if err != nil {
+		return nil, err
+	}
+	if n != nonceSize {
+		return nil, fmt.Errorf("unable to read sufficient random bytes")
+	}
+	cipherText := aead.Seal(result[nonceSize:nonceSize], result[:nonceSize], plaintext, s.dataCtx.AuthenticatedData())
+
+	return &service.EncryptResponse{
+		Ciphertext: result[:nonceSize+len(cipherText)],
+		KeyID:      s.keyID,
+		Annotations: map[string][]byte{
+			mockAnnotationKey: []byte("1"),
+		},
+	}, nil
+}
+
+func (s *pkcs11RemoteService) Decrypt(ctx context.Context, uid string, req *service.DecryptRequest) ([]byte, error) {
+	if len(req.Annotations) != 1 {
+		return nil, fmt.Errorf("invalid annotations")
+	}
+	if v, ok := req.Annotations[mockAnnotationKey]; !ok || string(v) != "1" {
+		return nil, fmt.Errorf("invalid version in annotations")
+	}
+
+	if req.KeyID != s.keyID {
+		return nil, fmt.Errorf("invalid keyID")
+	}
+
+	aead, err := s.key.NewGCM()
+	if err != nil {
+		return nil, err
+	}
+	nonceSize := aead.NonceSize()
+	data := req.Ciphertext
+	if len(data) < nonceSize {
+		return nil, fmt.Errorf("the stored data was shorter than the required size")
+	}
+
+	return aead.Open(nil, data[:nonceSize], data[nonceSize:], s.dataCtx.AuthenticatedData())
+}
+
+func (s *pkcs11RemoteService) Status(ctx context.Context) (*service.StatusResponse, error) {
```

Comment:
Any health check we can perform here?

Reply:
Looked at `FindKey` and it seems to get all the existing keys and then filters based on the `id` or `label`. IMO, this would be as expensive as `Encrypt`, `Decrypt`call.

Reply:
Okay we can leave this as-is.